src/mlpack/methods/range_search/rs_model.hpp

```diff
@@ -19,12 +19,212 @@
+
+  //! Bichromatic range search on the given RSType considering the leafSize.
+  template<typename RSType>
+  void SearchLeaf(RSType* rs) const;
+
+ public:
+  //! Alias template necessary for visual c++ compiler.
+  template<template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+ using RSTypeT = RSType<TreeType>;
+
+  //! Default Bichromatic range search on the given RSType instance.
+  template<template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  void operator()(RSTypeT<TreeType>* rs) const;
+
+  //! Bichromatic range search on the given RSType specialized for KDTrees.
+  void operator()(RSTypeT<tree::KDTree>* rs) const;
+
+  //! Bichromatic range search on the given RSType specialized for BallTrees.
+  void operator()(RSTypeT<tree::BallTree>* rs) const;
+
+  //! Bichromatic range search specialized for octrees.
+  void operator()(RSTypeT<tree::Octree>* rs) const;
+
+  //! Construct the BiSearchVisitor.
+  BiSearchVisitor(const arma::mat& querySet,
+                  const math::Range& range,
+                  std::vector<std::vector<size_t>>& neighbors,
+                  std::vector<std::vector<double>>& distances,
+                  const size_t leafSize
+                  );
+};
+
+/**
+ * TrainVisitor sets the reference set to a new reference set on the given
+ * RSType. We use template specialization to differentiate those tree types that
+ * accept leafSize as a parameter. In these cases, a reference tree with proper
+ * leafSize is built from the referenceSet.
+ */
+class TrainVisitor : public boost::static_visitor<void>
+{
+ private:
+  //! The reference set to use for training.
+  arma::mat&& referenceSet;
+  //! The leaf size, used only by BinarySpaceTree.
+  size_t leafSize;
+  //! Train on the given RsType considering the leafSize.
+  template<typename RSType>
+  void TrainLeaf(RSType* rs) const;
+
+ public:
+  //! Alias template necessary for visual c++ compiler.
+  template<template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  using RSTypeT = RSType<TreeType>;
+
+  //! Default Train on the given RSType instance.
+  template<template<typename TreeMetricType,
+                    typename TreeStatType,
+                    typename TreeMatType> class TreeType>
+  void operator()(RSTypeT<TreeType>* rs) const;
+
+  //! Train on the given RSType specialized for KDTrees.
+  void operator()(RSTypeT<tree::KDTree>* rs) const;
+
+  //! Train on the given RSType specialized for BallTrees.
+  void operator()(RSTypeT<tree::BallTree>* rs) const;
+
+  //! Train specialized for octrees.
+  void operator()(RSTypeT<tree::Octree>* rs) const;
+
+  //! Construct the TrainVisitor object with the given reference set, leafSize
+  //! for BinarySpaceTrees, and tau and rho for spill trees.
```

Comment:
Can you fix the comment here please?  tau and rho and spill trees aren't part of RangeSearch.